doc/sphinx/proofs/writing-proofs/reasoning-inductives.rst

```diff
@@ -473,82 +473,112 @@ Induction
 
-.. exn:: No primitive equality found.
-   :undocumented:
+We describe in this section some special purpose tactics dealing with
+equality of inductive sets or types. These tactics use the
+equality :g:`eq:forall (A:Type), A -> A -> Prop`, simply written with the infix
+symbol :g:`=`.
 
-.. exn:: Not a discriminable equality.
-   :undocumented:
+:tacn:`discriminate` and :tacn:`injection`
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+.. tacn:: discriminate {? @induction_arg }
+
+   .. todo PR Did I drop too much? (i.e. "The argument :n:`@one_term`
+      is assumed to be a proof of a statement of :n:`@term__1 = @term__2`"
+      How would that work with a non-trivial one_term?
+
+   Proves any goal using any hypothesis in the context in the form
+   :n:`@term__1 = @term__2` that
+   states an impossible structural equality for an inductive type.
+   For example, :g:`(S (S O)) = (S O)` is impossible.
+
+   The tactic traverses the normal forms [1]_ of the terms looking for
+   subterms :g:`u` and :g:`w` that are subterms
+   of the normal forms of :n:`@term__1` and :n:`@term__2`, respectively,
+   placed in the same positions and whose head symbols are
+   different constructors. If such subterms are present, the equality is
+   impossible and the current goal is completed.  Otherwise the tactic fails.
+
+   If no :n:`@induction_arg` is provided, the tactic examines each hypothesis
+   for impossible conditions.
+
+   .. exn:: No discriminable equalities.
+      :undocumented:
+
+   :n:`@natural` (in :n:`@induction_arg`)
+     Equivalent to :tacn:`intros` :n:`until @natural; discriminate @ident`,
+     where :n:`@ident` is the identifier for the last introduced hypothesis.
 
-.. tacv:: discriminate @natural
+   :n:`@ident`  (in :n:`@induction_arg`)
+     Equivalent to :n:`intros until @ident; discriminate @ident`.
 
-   This does the same thing as :n:`intros until @natural` followed by
-   :n:`discriminate @ident` where :n:`@ident` is the identifier for the last
-   introduced hypothesis.
+   :n:`with @bindings`  (in :n:`@induction_arg`)
+     Equivalent to :n:`discriminate @one_term` but uses the given
+     bindings to instantiate parameters or hypotheses of :n:`@one_term`.
 
-.. tacv:: discriminate @term with @bindings
+   .. note::
+      The syntax :n:`discriminate @ident` can be used to refer to a hypothesis
+      quantified in the goal. In this case, the quantified hypothesis whose name is
+      :n:`@ident` is first introduced in the local context using
+      :n:`intros until @ident`.
 
-   This does the same thing as :n:`discriminate @term` but using the given
-   bindings to instantiate parameters or hypotheses of :n:`@term`.
+   .. exn:: No primitive equality found.
+      :undocumented:
 
-.. tacv:: ediscriminate @natural
-          ediscriminate @term {? with @bindings}
+   .. exn:: Not a discriminable equality.
+      :undocumented:
+
+.. tacn:: ediscriminate {? @induction_arg }
+          ediscriminate {? @induction_arg }
```

Comment:
Why is this shown twice? There is no difference between the two lines AFAICT.

Also, please indent to make this appear as a variant.